Implement MSAA

VReaperV opened the pull request:
Implements #532.

Add `r_msaa`. When set to > 0, an MSAA FBO will be created with that many samples. This FBO will be used for rendering, other than when it requires sampling from current render/depth image. When such rendering is required the MSAA FBO will be blit into the main FBO and vice versa, to resolve the MSAA texture.

illwieckz commented:
Nice! Finally some antialiasing!

It looks like when MSAA is enabled the heathaze from back surface is applied on front surfaces too:
https://imgsli.com/MzMyODgx

illwieckz commented:
My previous comment was about screenshots from the original post.

When I try the branch and enable MSAA I get a black screen (AMD Radeon with Mesa radeonsi driver).

VReaperV commented:
> When I try the branch and enable MSAA I get a black screen (AMD Radeon with Mesa radeonsi driver).

That's interesting. Can you try with effects that use the main render target disabled, like heatHaze, bloom etc (or just use the lowest profile to make sure)? Also, does it work in the main menu?

slipher commented:
I get a completely black screen (other than 2D drawing) with `r_msaa` 4 or 16 on Spacetracks. Plat23, on the other hand, renders. I tested with Nvidia on Windows; default cvars except with reliefmapping enabled.

slipher commented:
Actually it's just black from whenever the 2nd map is loaded. Must be some cleanup bug

illwieckz commented:
Can this be rebased on current `master`? 🙂️

VReaperV commented:
> Can this be rebased on current `master`? 🙂️

Done... but I haven't tested it after rebase, so I don't know if it still works lol.

illwieckz commented:
It looks like the rebase went wrong! 😅️

```
/home/vsts/work/1/s/src/engine/renderer/tr_backend.cpp:3105:1: error: expected expression
<<<<<<< HEAD
^
```

slipher commented:
Works great on my favorite jaggies test case from Spacetracks!

Default:
![unvanquished-spacetracks-aliasing-jaggies](https://github.com/user-attachments/assets/f0d9ac63-7e8b-49d4-a6f5-8459e758d2ed)

With `r_msaa 1` - somehow a lot better already. Is that even supposed to be different from having it off?
![unvanquished-spacetracks-aliasing-jaggies](https://github.com/user-attachments/assets/05a1fa26-022b-465d-bf60-e1572dca643e)

`r_msaa 64`
![unvanquished-spacetracks-aliasing-jaggies](https://github.com/user-attachments/assets/9d35ac0d-87a9-4e40-9551-d76d1dd063fd)

illwieckz commented:
On my end it now works. I still get a black screen on AMD and Mesa when I use values higher than 8, but maybe we can cap that.

VReaperV commented:
> On my end it now works. I still get a black screen on AMD and Mesa when I use values higher than 8, but maybe we can cap that.

Does it say in the log that the samples requested were too high and it's falling back on some other value? According to https://opengl.gpuinfo.org/displaycapability.php?name=GL_MAX_COLOR_TEXTURE_SAMPLES, 8 is indeed the max for AMD, but the renderer should be automatically using that value if you specify a higher one.

illwieckz commented:
We probably need something similar to this:

```
	// HACK: We want to set the current value, not the latched value
	Cvar::ClearFlags("r_customwidth", CVAR_LATCH);
	Cvar::ClearFlags("r_customheight", CVAR_LATCH);
	Cvar_Set( "r_customwidth", va("%d", windowConfig.vidWidth ) );
	Cvar_Set( "r_customheight", va("%d", windowConfig.vidHeight ) );
	Cvar::AddFlags("r_customwidth", CVAR_LATCH);
	Cvar::AddFlags("r_customheight", CVAR_LATCH);
```

VReaperV commented:
Ah, yes. I'll add that a bit later.

slipher commented:
> We probably need something similar to this, but for `r_MSAA`:

Well that's only needed if you really want to use the cvar as a status display for the currently used value. Normally we try to avoid modifying cvars, and would store the current value in `glConfig.msaa` or something like that.

VReaperV commented:
That's an option too.

VReaperV commented on src/engine/renderer/tr_image.cpp:
```diff
@@ -1217,149 +1235,147 @@ void R_UploadImage( const char *name, const byte **dataArray, int numLayers, int
-	switch ( image->filterType )
-	{
-		case filterType_t::FT_DEFAULT:
+	if ( !samples ) {
```
I added a commit that does that.

slipher commented on src/engine/renderer/tr_fbo.cpp:
```diff
@@ -356,24 +357,31 @@ void R_AttachFBOTexturePackedDepthStencil( int texId )
-	if ( glState.currentFBO != fbo )
-	{
-		GL_fboShim.glBindFramebuffer( GL_FRAMEBUFFER, fbo->frameBuffer );
+	if ( glState.currentFBO != fbo ) {
```
How about don't read or update `currentFBO` is the `target` is not the default one? Then the `glState.currentFBO = nullptr` stuff is not needed

VReaperV replied:
Done.

slipher commented on src/engine/renderer/tr_fbo.cpp:
```diff
@@ -430,6 +438,17 @@ void R_InitFBOs()
+		R_AttachFBOTexture2D( GL_TEXTURE_2D_MULTISAMPLE, tr.currentRenderImageMSAA->texnum, 0 );
+		GL_CheckErrors();
+		R_AttachFBOTexturePackedDepthStencilMSAA( tr.currentDepthImageMSAA->texnum );
+		GL_CheckErrors();
```
Do we want to keep CheckErrors after every line?

slipher commented:
Also depth fade isn't working. It looks the same as if depth fade were disabled.

![unvanquished-vega-depthfade-glitch-decal](https://github.com/user-attachments/assets/b4c20695-5d44-45f8-951b-0fc3a482edb0)

slipher commented:
Also I confirm what illwieckz observed long ago about heat haze penetrating opaque surfaces. Try the following commands and you will see heat haze on the other side of a door being rendered.

```
devmap chasm
setviewpos -634 67 79 -178 4
testshader gfx/weapons/flamer/haze
```

![unvanquished-chasm-heathaze-behind-door](https://github.com/user-attachments/assets/c184787d-15c6-4b29-a8ed-319160160bcb)

VReaperV commented:
> Also depth fade isn't working. It looks the same as if depth fade were disabled.
> 
> ![unvanquished-vega-depthfade-glitch-decal](https://private-user-images.githubusercontent.com/7809431/485920348-b4c20695-5d44-45f8-951b-0fc3a482edb0.jpg?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTcwNzAwMjYsIm5iZiI6MTc1NzA2OTcyNiwicGF0aCI6Ii83ODA5NDMxLzQ4NTkyMDM0OC1iNGMyMDY5NS01ZDQ0LTQ1ZjgtOTUxYi0wZmMzYTQ4MmVkYjAuanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDkwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA5MDVUMTA1NTI2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZGQwMmQ1ZWZjZDIzNzYyZTllNTkxMjU2OWM5NzA5YmZmZDVlZTkxM2U4NTEyZTQ5NjEwNDM4YWI0YzYxY2U3ZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.TES_pdkjhF-X-GOGIOW2-TjvvDMi7nWRSIwdy1U2r1U)

Fixed.



> Also I confirm what illwieckz observed long ago about heat haze penetrating opaque surfaces. Try the following commands and you will see heat haze on the other side of a door being rendered.
> 
> ```
> devmap chasm
> setviewpos -634 67 79 -178 4
> testshader gfx/weapons/flamer/haze
> ```
> 
> ![unvanquished-chasm-heathaze-behind-door](https://private-user-images.githubusercontent.com/7809431/485922666-c184787d-15c6-4b29-a8ed-319160160bcb.jpg?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTcwNzAwMjYsIm5iZiI6MTc1NzA2OTcyNiwicGF0aCI6Ii83ODA5NDMxLzQ4NTkyMjY2Ni1jMTg0Nzg3ZC0xNWM2LTRiMjktYThlZC0zMTkxNjAxNjBiY2IuanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDkwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA5MDVUMTA1NTI2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDlkNzgwNWY5NmVlNzQzMTYyNTZiOWNiZWQ1ODlhMmQ2OTc4ZTU5Nzc0MmVlNWQ1N2MxNTRmMGU4MTMxZGM1NyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.1AIfDYJMpwOvSx3hlRWJf8tJaHwEYIbZyefZSsZo9dc)

Also fixed now.

VReaperV commented:
> There's a bug - dynamic lights are not rendered when MSAA is enabled.

Fixed this too.

VReaperV commented:
> In the latest version, some transparent surfaces are no longer rendered. For example in the Vega scene below, the cone under the globe and the alpha-tested grates are missing. Also fogs are gone.
> 
> ![unvanquished_2025-09-05_130336_000](https://private-user-images.githubusercontent.com/7809431/486226510-fa8e166d-126f-4a07-bfc3-9c4ebe304722.jpg?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjY3MzA2MTQsIm5iZiI6MTc2NjczMDMxNCwicGF0aCI6Ii83ODA5NDMxLzQ4NjIyNjUxMC1mYThlMTY2ZC0xMjZmLTRhMDctYmZjMy05YzRlYmUzMDQ3MjIuanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTIyNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTEyMjZUMDYyNTE0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTNhNDZkZDA2ZjM5NTc2ZmUwMjYzNjM5ZmNjYzg2YTA4YjM0MjhlNTU4OGFhYzFjNmFkZWI1MjhlNjU1MWRkOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Y8CoUm9cfcMLlUSmk633H7kMbNNJbWwMGLQxjL9Jv_s)

Both fixed now.

slipher commented:
Turning on `r_msaa` causes tone mapping to be disabled in my tests. This is the same issue as I saw in https://github.com/DaemonEngine/Daemon/pull/1684#issuecomment-3624398313 (but this time I took screenshots with `r_toneMapping 0` to confirm that this is what is happening). I can't reproduce it manually, only in scripted tests, so I need to do more investigation on how to reproduce it.

VReaperV commented:
By numerical imprecision I mean using the same vertices for the two draws producing slightly different results due to the rasterisation specifics.

VReaperV commented:
> * SSAO does not work.

Fixed.

VReaperV commented:
> Turning on `r_msaa` causes tone mapping to be disabled in my tests. This is the same issue as I saw in [#1684 (comment)](https://github.com/DaemonEngine/Daemon/pull/1684#issuecomment-3624398313) (but this time I took screenshots with `r_toneMapping 0` to confirm that this is what is happening). I can't reproduce it manually, only in scripted tests, so I need to do more investigation on how to reproduce it.
> Also color grading is missing in my tests so it's like the whole cameraEffects shader is skipped?

I'm unable to reproduce this on NV and on Intel Arc (both with proprietary drivers).

slipher commented:
Figured out what the tonemapping issue is. Nothing to do with my scripts; the problem is that screenshots are not the same as what you seen is on-screen. It seems to show the 3D drawing framebuffer before any color effects or 2D drawing.

slipher commented:
Yeah looks like RB_ReadPixels doesn't bother to select which framebuffer to read from. So it depends on the correct one being active by accident and this could depend on the sequence of `RenderCommand`s.

slipher commented on src/engine/renderer/tr_backend.cpp:
```diff
@@ -199,6 +199,47 @@ GLuint64 GL_BindToTMU( int unit, image_t *image )
+	glBlitFramebuffer( 0, 0, fbo->width, fbo->height, 0, 0, tr.msaaFBO->width, tr.msaaFBO->height,
+		mask, GL_NEAREST );
+
+	R_BindFBO( GL_DRAW_FRAMEBUFFER, fbo );
```
This `R_BindFBO` call seems pointless since the calling function immediately does its own `R_BindFBO`.

VReaperV replied:
Changed to just bind `GL_READ_FRAMEBUFFER`.

slipher commented on src/engine/renderer/tr_shade.cpp:
```diff
@@ -880,11 +887,21 @@ void Render_generic3D( shaderStage_t *pStage )
-	if ( needDepthMap )
+	const bool needMSAATransion = needDepthMap && backEnd.dirtyDepthBuffer;
+
+	if ( needDepthMap && backEnd.dirtyDepthBuffer && glConfig.textureBarrierAvailable )
```
I suggest writing the code like this:
```
if ( needDepthMap )
{
    RB_PrepareForSamplingDepthMap();
    
    if ( glConfig.MSAA && backEnd.dirtyDepthBuffer )
    {
        TransitionMSAAToMain( GL_DEPTH_BUFFER_BIT );
        R_BindFBO( tr.msaaFBO );
    }
}
``` 

Since RB_PrepareForSamplingDepthMap already checks the texture barrier available and dirty depth buffer conditions. And the following TransitionMSAAToMain and R_BindFBO calls always occur together, so the intent is clearer if they are in the same if block.

VReaperV replied:
Oh, yea, `RB_PrepareForSamplingDepthMap()` must've been changed since this was written. Fixed now.

slipher commented on src/engine/renderer/tr_image.cpp:
```diff
@@ -1150,12 +1159,21 @@ void R_UploadImage( const char *name, const byte **dataArray, int numLayers, int
+						GL_TexImage2D( target, 0, internalFormat, scaledWidth, scaledHeight, 0, format, GL_UNSIGNED_BYTE, scaledBuffer, isSRGB );
+					}
 				}
+				GL_CheckErrors();
```
GL_CheckErrors not needed here since it is done after the switch statement.

VReaperV replied:
Where after the switch statement?

slipher replied:
Maybe I should say after the outermost if/else. Right before the new location of `switch ( image->internalFormat )`

VReaperV commented on src/engine/renderer/tr_backend.cpp:
```diff
@@ -199,6 +199,47 @@ GLuint64 GL_BindToTMU( int unit, image_t *image )
+		mask, GL_NEAREST );
+
+	R_BindFBO( GL_READ_FRAMEBUFFER, fbo );
+	glState.currentFBO = fbo;
```
Changed to just bind `GL_READ_FRAMEBUFFER`.

VReaperV commented:
> Unfortunately I found another bug when trying a different graphics preset... when MSAA is on and SSAO is off, depth fade doesn't work (no alpha modification occurs). I reproduced it with Nvidia so you should hopefully be able to see it.

Fixed I think. Also added the missing MSAA resolve to global fog.
